test/test_numerics.py

```diff
@@ -1,14 +1,46 @@
 # Copyright (c) 2023 Graphcore Ltd. All rights reserved.
 import jax.numpy as jnp
+import pytest
 from numpy.testing import assert_allclose
 
-from pyscf_ipu.experimental.integrals import factorial_fori, factorial_gamma
-from pyscf_ipu.experimental.numerics import compare_fp32_to_fp64
+from pyscf_ipu.experimental.special import (
+    binom_beta,
+    binom_fori,
+    binom_lookup,
+    factorial2_fori,
+    factorial2_lookup,
+    factorial_fori,
+    factorial_gamma,
+    factorial_lookup,
+)
 
 
 def test_factorial():
-    n = 16
-    x = jnp.arange(n, dtype=jnp.float32)
-    y_fori = compare_fp32_to_fp64(factorial_fori)(x, n)
-    y_gamma = compare_fp32_to_fp64(factorial_gamma)(x)
-    assert_allclose(y_fori, y_gamma, 1e-2)
+    x = jnp.array([1, 2, 3, 4, 5, 6, 7, 8])
+    expect = jnp.array([1, 2, 6, 24, 120, 720, 5040, 40320])
+    assert_allclose(factorial_fori(x, x[-1]), expect)
+    assert_allclose(factorial_lookup(x, x[-1]), expect)
+    assert_allclose(factorial_gamma(x), expect)
+
+
+def test_factorial2():
+    x = jnp.array([1, 2, 3, 4, 5, 6, 7, 8])
+    expect = jnp.array([1, 2, 3, 8, 15, 48, 105, 384])
+    assert_allclose(factorial2_fori(x), expect)
+    assert_allclose(factorial2_fori(0), 1)
+
+    assert_allclose(factorial2_lookup(x), expect)
+    assert_allclose(factorial2_lookup(0), 1)
+
+
+@pytest.mark.parametrize("binom_func", [binom_beta, binom_fori, binom_lookup])
+def test_binom(binom_func):
+    x = jnp.array([4, 4, 4, 4])
```

Comment:
Maybe test some edge cases too (1,1) (3,3) (3,0) (0,-1) etc

Reply:
good idea, I think $3 \choose 3$ and $3 \choose 0$ should already be covered but I've added some more cases involving ones, zeros, and minus ones.